File: AIS3-Pre-Exam-2021/Encrypted-Lost-Flag/src/solve.c
#include "matrix.h"
#include "config.c"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

int main() {
    // flag_enc * key = flag
    // banner_enc * key = banner
    // key = banner_enc_inv * banner
    Matrix* banner_m = InitMatrixString(size, banner);
    Matrix* banner_enc_m = InitMatrixString(size, banner_enc);
    Matrix* banner_enc_inv_m = MatrixInverse(banner_enc_m);
    PrintMatrixInlineHex(banner_enc_inv_m);
    Matrix* key_m = MatrixMultipy(banner_enc_inv_m, banner_m);
    Matrix* flag_enc_m = InitMatrixString(size, flag_enc);
    Matrix* flag_m = MatrixMultipy(flag_enc_m, key_m);
    puts(GetStr(flag_m));
}

File: AIS3-Pre-Exam-2021/Encrypted-Lost-Flag/src/solve.py
from sage.all import *
from os import path
config = open(path.join(path.dirname(__file__),'config.c')).read()
for line in config.strip().split('\n'):
    x,y = line.split(' = ')
    n = x.split()[-1]
    if '"' in y: y = 'b'+y
    exec(f'{n} = {y}')

mod = 0x101
s2m = lambda s: matrix(GF(mod), size, size, s)

banner_m     = s2m(banner)
banner_enc_m = s2m(banner_enc)
flag_enc_m   = s2m(flag_enc)

flag_m = flag_enc_m * banner_enc_m.inverse() * banner_m
flag = ''.join(map(chr,vector(flag_m)))
print(flag)
